syntax = "proto3";

package jcarbon.service;

option java_multiple_files = true;
option java_package = "jcarbon.service";

service JCarbonService {
  rpc Start (StartRequest) returns (StartResponse) {}

  rpc Stop (StopRequest) returns (StopResponse) {}

  rpc Dump (DumpRequest) returns (DumpResponse) {}

  rpc Read (ReadRequest) returns (ReadResponse) {}

  rpc Purge (PurgeRequest) returns (PurgeResponse) {}
}

message StartRequest {
  optional uint64 process_id = 1;
  optional uint32 period_millis = 2;
}

message StartResponse {
  optional string response = 1;
}

message StopRequest {
  optional uint64 process_id = 1;
}

message StopResponse {
  optional string response = 1;
}

message DumpRequest {
  optional uint64 process_id = 1;
  optional string output_path = 2;
  repeated string signals = 3;
}

message DumpResponse {
  optional string response = 1;
}

message ReadRequest {
  optional uint64 process_id = 1;
  repeated string signals = 2;
}

message ReadResponse {
  optional JCarbonReport report = 1;
}

message PurgeRequest {}

message PurgeResponse {}

message JCarbonReport {
  repeated JCarbonSignal signal = 1;
}

message JCarbonSignal {
  optional string signal_name = 1;
  repeated Signal signal = 2;
}

message Signal {
  optional Timestamp start = 1;
  optional Timestamp end = 2;
  optional string component = 3;
  optional string unit = 4;
  repeated Data data = 5;

  message Timestamp {
    optional uint64 secs = 1;
    optional uint64 nanos = 2;
  }

  message Data {
    optional string component = 1;
    optional double value = 2;
  }
}